# Семинар 3 (2.02)
[Содержание](https://drive.google.com/file/d/1H1dBsZiSSBKNATe3I3QWZL4cOHkOICxp/view)

Генотип - состояние двух его хромосом в конкретном месте.

Локус - место на хромосоме.

Аллель - состояние локуса на 1 хромосоме. В локусе 2 аллеля

У растений возможно дублирование генома, хромосом у них может быть много. Два аллеля - это частный случай.

МЫ предполагаем у людей новый вид болезни. Мы хотим секвенировать геномы и найти различия между здоровыми и больными людьми. Разберем примеры.

## Задание 1
Ставим геномы рядом

Человек 1

ATCAATTAATGTATTAGCA

ATCAATTAATGTATTAGCA

Человек 2

ATCAATTAATGTATTAGCA

TTCAATTAATGAATTAGCA

Человек 3

ATCAATAAATCTATTAGCA

TTCAATTAATGTATTAGCA

Человек 4

ATCAATAAATGTATTAGCA

ATCAATTAATGTATTAGCA

1.  Геномы выравниваются по референсному. Теперь считаем, что позиции геномов одинаковы относительно начальной последовательности (нет сдвигов генома, иначе позиции не соответствуют друг другу и они не гомологичны)

2.  Отмечаем локусы со снипами. 

Человек 1

ATCAATTAATGTATTAGCA

ATCAATTAATGTATTAGCA

Человек 2

ATCAATTAATGTATTAGCA

**T**TCAATTAATG**A**ATTAGCA

Человек 3

ATCAAT**A**AAT**C**TATTAGCA

**T**TCAATTAATGTATTAGCA

Человек 4

ATCAAT**A**AATGTATTAGCA

ATCAATTAATGTATTAGCA

(Выведили минорные аллели)

Эти 4 позиции - полиморфизмы

Добавляем человека:

Человек 5

ATCAATTAATGTATT**T**GCA

ATCAAT**A**AAT**C**TATTAGC**T**

Появились новые снипы. Получили 2 новых полиморфизма

3.   В локусе 2 аллеля. Частота минорной аллели - отношение количества минорных к общему. Больше 0,5 - мажорный, меньше - минорный. Минорность говорит только о частоте в нашей выборки, без какого-то признака мы не можем говорить о чем-то еще. MAF - minor allele frequency 

     MAF1 = 0,2 (первый полиморфизм)

     MAF(1, A->T) = 0,2 (то же самое, только указали позицию и конкретную замену)

     MAF2 = 0,3 (= MAF(7, T-> A))

     MAF3 = 0,2

     MAF4 = 0,1

     MAF5 = 0,1

     MAF6 = 0,1

Мы не задавали референс в нашей выборки, но можем его задать из тех аллелей, которые встречаются чаще всего. Частоты вообще говоря считаются от нее. В dbSNP слева указан референс, справа - то, что мы ей дали.

ref

ATCAATTAATGTATTAGCA

Референс может отражать малочисленную группу

Смотрим полиморфизмы в dbSNP - в поиске пишем tp53, берем верхний (ФОТО). Слева - референс, справа - возможные полиморфизмы. В данном случае референс минорный (ФОТО)

5.   Смотрим человека 3. енотип можно представить в виде 0(нет полиморфизма), 1(есть полиморфизм). Это стандартное обозначение, которое используется в файлах для генотипирования. Пишется 0/0, 0/1 и т.п в колонке, левые числа отвечают за одну хромосому, правые - за другую. В такой записи вводятся колонки, в которых указана замена


Смотрим на такую запись (ФОТО).

ID - если это полиморфизм, указывается номер из dbSNP

Ref, Alt - референс и альтернатива

На вики это [Variant Call Fromat](https://en.wikipedia.org/wiki/Variant_Call_Format)

Сделаем такую запись для человека 3

Человек 3

ATCAAT**A**AAT**C**TATTAGCA

**T**TCAATTAATGTATTAGCA

1    A T 0/1

7    0/1

11   0/1

16   0/0

19   0/0

Остальные позиции без полиморфизмов, там везде было бы 0/0

Можем считать корреляцию между локусами или полиморфизмами (что и сделаем):

1. 00 01 01 00 00 

2. 00 00 01 01 01

3. 00 00 01 00 01

4. 00 01 00 00 00 

5. 00 00 00 00 01

6. 00 00 00 00 01

У нас хромосомы неотличимы, поэтому 01 - 10 для нас неважно, для единообразия пишем 01

Считаем коэффициент корреляции (ФОТО)

Коэф. корреляции - можно из курса матстата узнать. У нас есть две случайные величины, и хотим увидеть, насколько графики распределений схожи. Если они схожи, то коэффициент положительный, если графики противоположны - отрицательный и ближе к 1, если они никак не связаны - то около 0. Когда случайные величины независимы, он равен 0, иначе он не равен 0. 


Хотим узнать, связаны ли полиморфизмы. 5 и 6 выглядят одинаково -  возможно, они произошли одновременно. Так мы можем строить гипотезы о возникновении заболеваний. Пусть есть 2 полиморфизма в разных генах. Если они появляются одновременно, то возможно они появляются вместе (вероятность того, что это совпадение, сохраняется).

Создадим таблицу, как показано выше, для 1 и 2 полиморфизма (ФОТО)

Подставляем значения в формулу (ФОТО)

Похоже, что 1 и 2 снипы являются независимыми. Для 5 и 6 снипа b = 0, c = 0 (ФОТО). Величины скоррелированы. Коррелированность не равна зависимости (ФОТО) 

Рассматриваем пример (ФОТО). Case - больной, Control - здоровый. Смотрим число мажорных и минорных болезней. Хотим узнать, связаны они или нет. Используем Хи-квадрат (тест Фишера). Идея: рассматривая распределение аллелей и признака, мы принимаем их как случайные величины, следовательно, значения в таблице были получены случайно с какой-то вероятностью. Мы хотим узнать, насколько наблюдаемая нами матрица маловероятна (сравниваем полученные частоты с ожидаемыми, если отклонение большое, то мы считаем, что это не случайность).Так находим значение критерия хи-квадрат, и если он больше какого-то порога, то мы считаем, что наша выборка является неслучайной.

Заполним таблицу, указав общее количество и частоту (ФОТО)

Пусть в коробке лежат черные и белые предмеы, шарики и кубики. Вероятность черного 0,5, кубика 0,5. какова вероятность вытащить черный куб? 0,25. Так как признаки независимы. Также если признаки независимы, то мы можем их перемножить. Смотрим на частоту контроля и минорности и их перемножить, так как они независимы как признаки. Это будет ожидаемая вероятность. Домножим на общее число - получим ожидаемое число (ФОТО). 

Как находить p-value для таблиц? Тест Фишера и хи-квадрат. Минус теста Фишера - долго считается на больших данных, там в формуле есть факториалы. Поэтому точный тест Фишера на маленьких, в то время как хи-квадрат лучше использовать на больших, на маленьких у него проблемы.

In [ ]:
from scipy.stats import chi2_contingency
stat, p, dof, expected =chi2_contingency([[4, 102],[2, 99]],
correction=False)
print(p)
#точный тест Фишера
from scipy import stats
table=[[2,99],[4,102]]
oddsratio, pvalue=stats.fisher_exact(table)
print(pvalue)
print(oddsratio)

0.4420270227417106
0.6834472680224617
0.5151515151515151


Можно забить это в калькуляторе, а можно так. У нас p-value получилось 0.4, для точного теста 0.68. То есть эти признаки не являются связанными (p-value большое), то есть локус не является статистически значимым.

GWAS - полногеномное исследование по ассоциациям. Там данные по различным признакам и генотипам.

Done.